### PR TITLE
SOLR-15252 Solr should log WARN log when a query requests huge rows number

### PR DESCRIPTION
**Pull request by janhoy**
See https://issues.apache.org/jira/browse/SOLR-15252

~Please consider only changes in QueryComponent.java, as the other changes is to get Solr compile (#77)~

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java [hunk 1]
```diff
@@ -236,6 +240,21 @@ public void prepare(ResponseBuilder rb) throws IOException
...
     }
+    if(rb.getSortSpec().getCount() > ROWS_WARN_THRESHOLD && shouldLogPeriodically("rowsWarn", 60)) {
+      log.warn("Very high 'rows' parameter detected. This may lead to performance- and memory problems. " +
+          "Consider pagination, see https://solr.apache.org/guide/pagination-of-results.html. " +
+          "This warning will mute for 60s.");
+    }
+  }
+
+  // Decides whether to log again based on key and interval
+  private static boolean shouldLogPeriodically(String key, int intervalSeconds) {
```
Comment (janhoy) [inline solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java:251]: I know this is a bit of an overkill for just one use. Are you aware of similar logic elsewhere that we could use instead? Or some Util class where this static method could live?

File: solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java [hunk 2]
```diff
@@ -236,6 +240,21 @@ public void prepare(ResponseBuilder rb) throws IOException
     if (rb.getSortSpec().getOffset() < 0) {
       throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "'start' parameter cannot be negative");
     }
+    if(rb.getSortSpec().getCount() > ROWS_WARN_THRESHOLD && shouldLogPeriodically("rowsWarn", 60)) {
+      log.warn("Very high 'rows' parameter detected. This may lead to performance- and memory problems. " +
+          "Consider pagination, see https://solr.apache.org/guide/pagination-of-results.html. " +
+          "This warning will mute for 60s.");
```
Comment (janhoy) [inline solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java:246]: Should the default mute be 3600s instead? 24 WARN logs per day should be enough to draw attention to this issue in most systems.

**Comment by cpoerschke** [comment]
Any thoughts on also mentioning this in the Solr Ref guide documentation e.g. https://github.com/apache/solr/blob/main/solr/solr-ref-guide/src/common-query-parameters.adoc#rows-parameter or maybe it's hopefully too uncommon to mention in the common parameters context?

**Comment by cpoerschke** [comment]
Any thoughts on also mentioning this in the Solr Ref guide documentation e.g. https://github.com/apache/solr/blob/main/solr/solr-ref-guide/src/common-query-parameters.adoc#rows-parameter or maybe it's hopefully too uncommon to mention in the common parameters context?

File: solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java (hunk 2 above)
Comment (madrob) [reply solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java:246]: I think maybe 300s is a better compromise. I'd probably miss something that logs hourly, based on my experiences with heavily used clusters - there are plenty of WARN messages already about slow queries, leader elections, other "normal" operations in a large cluster.

Comment (cpoerschke) [reply solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java:246]: How about making the logging interval (and the threshold) configurable? I've put up a possible approach in https://github.com/cominvent/solr/compare/solr15252-log-high-rows...cpoerschke:solr15252-log-high-rows and the idea would be that both could be made default (or invariant) parameters on the request handler. Admittedly `rows.warnThresholdLogIntervalSeconds` is a bit of a long name though.

**Pull request closed without merging**